OnPod/Node/Test/EventScheduler/main.cpp

```diff
@@ -0,0 +1,253 @@
+    TEST_ASSERT_EQUAL_UINT16(2, test_callback_run_count);
+    resetTestCallback();
+}
+
+void callFunctionEveryForever_delayPassedThousandTimes_functionCalledThousandTimes(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionEvery(1000, testCallback, 0);
+    for (uint16_t i = 1; i <= 1000; i++)
+    {
+        event_scheduler.updateEvents(i*1001);
+    }
+    TEST_ASSERT_EQUAL_UINT16(1000, test_callback_run_count);
+    resetTestCallback();
+}
+
+void callFunctionEveryForever_delayNotPassed_functionNotCalled(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionEvery(1000, testCallback, 0);
+    event_scheduler.updateEvents(999);
+    TEST_ASSERT_EQUAL_UINT16(0, test_callback_run_count);
+    resetTestCallback();
+}
+
+void callFunctionEveryForever_delayNotPassedThenPassed_functionCalledOnce(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionEvery(1000, testCallback, 0);
+    event_scheduler.updateEvents(999);
+    event_scheduler.updateEvents(1001);
+    TEST_ASSERT_EQUAL_UINT16(1, test_callback_run_count);
+    resetTestCallback();
+}
+
+void callFunctionEveryForever_delayExactlyReached_functionCalledOnce(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionEvery(1000, testCallback, 0);
+    event_scheduler.updateEvents(1000);
+    TEST_ASSERT_EQUAL_UINT16(1, test_callback_run_count);
+    resetTestCallback();
+}
+
+// callFunctionEvery (finite)
+void callFunctionEveryFinite_delayNotPassed_functionNotCalled(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionEvery(1000, testCallback, 0, 2);
+    event_scheduler.updateEvents(999);
+    TEST_ASSERT_EQUAL_UINT16(0, test_callback_run_count);
+    resetTestCallback();
+}
+
+void callFunctionEveryFinite_delayPassed_functionCalledOnce(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionEvery(1000, testCallback, 0, 2);
+    event_scheduler.updateEvents(1001);
+    TEST_ASSERT_EQUAL_UINT16(1, test_callback_run_count);
+    resetTestCallback();
+}
+
+void callFunctionEveryFinite_delayPassedTwice_functionCalledTwice(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionEvery(1000, testCallback, 0, 2);
+    event_scheduler.updateEvents(1001);
+    event_scheduler.updateEvents(2002);
+    TEST_ASSERT_EQUAL_UINT16(2, test_callback_run_count);
+    resetTestCallback();
+}
+
+void callFunctionEveryFinite_delayPassedThreeTimes_functionCalledTwice(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionEvery(1000, testCallback, 0, 2);
+    event_scheduler.updateEvents(1001);
+    event_scheduler.updateEvents(2002);
+    event_scheduler.updateEvents(3003);
+    TEST_ASSERT_EQUAL_UINT16(2, test_callback_run_count);
+    resetTestCallback();
+}
```

Comment:
Maybe repeat this test except create a new event and ensure the event_id is 0 and not 1? To make sure the old complete event is properly cleared for use. Just to check off all the boxes?

Reply:
I'm kinda confused about what you mean here

Reply:
I just mean maybe test that finite events get cleaned up after they "expire", could be too explicit of a test though.

Reply:
store the id returned by event 1000, let it expire. Create event 2000 and the id should be the same as 1000, both being 0. this ensures 1000 was cleaned up and replaced properly.

Additionally, I would fill with 51 events to ensure the -1 test recommended in coltons other comment works, have 1 expire then try to add a 52 which should succed. This to check an edge case around the max event buffer size.